llvm/lib/CodeGen/MachineCycleAnalysis.cpp

```diff
@@ -54,43 +54,63 @@ void MachineCycleInfoWrapperPass::releaseMemory() {
   F = nullptr;
 }
 
+AnalysisKey MachineCycleAnalysis::Key;
+
+MachineCycleInfo
+MachineCycleAnalysis::run(MachineFunction &MF,
+                          MachineFunctionAnalysisManager &MFAM) {
+  MachineCycleInfo MCI;
+  MCI.compute(MF);
+  return MCI;
+}
+
 namespace {
-class MachineCycleInfoPrinterPass : public MachineFunctionPass {
+class MachineCycleInfoPrinterLegacy : public MachineFunctionPass {
 public:
   static char ID;
 
-  MachineCycleInfoPrinterPass();
+  MachineCycleInfoPrinterLegacy();
 
   bool runOnMachineFunction(MachineFunction &F) override;
   void getAnalysisUsage(AnalysisUsage &AU) const override;
 };
 } // namespace
 
-char MachineCycleInfoPrinterPass::ID = 0;
+char MachineCycleInfoPrinterLegacy::ID = 0;
 
-MachineCycleInfoPrinterPass::MachineCycleInfoPrinterPass()
+MachineCycleInfoPrinterLegacy::MachineCycleInfoPrinterLegacy()
     : MachineFunctionPass(ID) {
-  initializeMachineCycleInfoPrinterPassPass(*PassRegistry::getPassRegistry());
+  initializeMachineCycleInfoPrinterLegacyPass(*PassRegistry::getPassRegistry());
 }
 
-INITIALIZE_PASS_BEGIN(MachineCycleInfoPrinterPass, "print-machine-cycles",
+INITIALIZE_PASS_BEGIN(MachineCycleInfoPrinterLegacy, "print-machine-cycles",
                       "Print Machine Cycle Info Analysis", true, true)
 INITIALIZE_PASS_DEPENDENCY(MachineCycleInfoWrapperPass)
-INITIALIZE_PASS_END(MachineCycleInfoPrinterPass, "print-machine-cycles",
+INITIALIZE_PASS_END(MachineCycleInfoPrinterLegacy, "print-machine-cycles",
                     "Print Machine Cycle Info Analysis", true, true)
 
-void MachineCycleInfoPrinterPass::getAnalysisUsage(AnalysisUsage &AU) const {
+void MachineCycleInfoPrinterLegacy::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.setPreservesAll();
   AU.addRequired<MachineCycleInfoWrapperPass>();
   MachineFunctionPass::getAnalysisUsage(AU);
 }
 
-bool MachineCycleInfoPrinterPass::runOnMachineFunction(MachineFunction &F) {
+bool MachineCycleInfoPrinterLegacy::runOnMachineFunction(MachineFunction &F) {
   auto &CI = getAnalysis<MachineCycleInfoWrapperPass>();
   CI.print(errs());
   return false;
 }
 
+PreservedAnalyses
+MachineCycleInfoPrinterPass::run(MachineFunction &MF,
+                                 MachineFunctionAnalysisManager &MFAM) {
+  OS << "MachineCycleInfo for function: " << MF.getName() << "\n";
```

Comment:
```suggestion
  OS << "MachineCycleInfo for function: " << MF.getName() << '\n';
```

Do we really need this print at all? Do other passes have it? It will not handle anonymous functions correctly as-is.

Reply:
I think we should rip all of this out. This can be done generically outside the pass if it's useful

Reply:
There is one test that checks for cycle info using this printer pass, so I'm not sure.
One alternative to `print<machine-cycles>` is printing under `debug` and running a `require<machine-cycles>` pass.

Reply:
I mean the pre-printing of the name part, not the whole print

Reply:
Still not convinced, two things:
- Leaving out the MF name pre print would lose the printing information available through minimal effort.

Tests usually do this:
```
; CHECK: SomeAnalysis for function: someFunction
; CHECK-NEXT:  SomeAnalysis for BB %0 is:
...
; CHECK: SomeAnalysis for function: otherFunction
...
```

- For anonymous functions, this prints an empty string. Is any special handling needed?

Reply:
> Still not convinced, two things:
> 
> * Leaving out the MF name pre print would lose the printing information available through minimal effort.
> 
> Tests usually do this:


This should not be the responsibility of every single pass.  The pass manager itself can directly print the name of the pass without requiring this boilerplate everywhere

> 
> ```
> ; CHECK: SomeAnalysis for function: someFunction
> ; CHECK-NEXT:  SomeAnalysis for BB %0 is:
> ...
> ; CHECK: SomeAnalysis for function: otherFunction
> ...
> ```
> 
>  * For anonymous functions, this prints an empty string. Is any special handling needed?

Yes, that's the problem. printAsOperand handles this case

Reply:
Alright, got it.
-print-{before|after} options already print MF info so we can remove it from here.